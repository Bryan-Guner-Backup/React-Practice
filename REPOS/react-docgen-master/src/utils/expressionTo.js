/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

/*eslint no-loop-func: 0, no-use-before-define: 0*/

import { namedTypes as t } from 'ast-types';
import resolveToValue from './resolveToValue';
import type { Importer } from '../types';

/**
 * Splits a MemberExpression or CallExpression into parts.
 * E.g. foo.bar.baz becomes ['foo', 'bar', 'baz']
 */
function toArray(path: NodePath, importer: Importer): Array<string> {
  const parts = [path];
  let result = [];

  while (parts.length > 0) {
    path = parts.shift();
    const node = path.node;
    if (t.CallExpression.check(node)) {
      parts.push(path.get('callee'));
      continue;
    } else if (t.MemberExpression.check(node)) {
      parts.push(path.get('object'));
      if (node.computed) {
        const resolvedPath = resolveToValue(path.get('property'), importer);
        if (resolvedPath !== undefined) {
          result = result.concat(toArray(resolvedPath, importer));
        } else {
          result.push('<computed>');
        }
      } else {
        result.push(node.property.name);
      }
      continue;
    } else if (t.Identifier.check(node)) {
      result.push(node.name);
      continue;
    } else if (t.Literal.check(node)) {
      result.push(node.raw);
      continue;
    } else if (t.ThisExpression.check(node)) {
      result.push('this');
      continue;
    } else if (t.ObjectExpression.check(node)) {
      const properties = path.get('properties').map(function (property) {
        return (
          toString(property.get('key'), importer) +
          ': ' +
          toString(property.get('value'), importer)
        );
      });
      result.push('{' + properties.join(', ') + '}');
      continue;
    } else if (t.ArrayExpression.check(node)) {
      result.push(
        '[' +
          path
            .get('elements')
            .map(function (el) {
              return toString(el, importer);
            })
            .join(', ') +
          ']',
      );
      continue;
    }
  }

  return result.reverse();
}

/**
 * Creates a string representation of a member expression.
 */
function toString(path: NodePath, importer: Importer): string {
  return toArray(path, importer).join('.');
}

export { toString as String, toArray as Array };
