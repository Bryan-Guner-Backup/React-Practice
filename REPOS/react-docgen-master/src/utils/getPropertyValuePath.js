/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import getPropertyName from './getPropertyName';
import type { Importer } from '../types';

/**
 * Given an ObjectExpression, this function returns the path of the value of
 * the property with name `propertyName`.
 */
export default function getPropertyValuePath(
  path: NodePath,
  propertyName: string,
  importer: Importer,
): ?NodePath {
  t.ObjectExpression.assert(path.node);

  return path
    .get('properties')
    .filter(
      propertyPath => getPropertyName(propertyPath, importer) === propertyName,
    )
    .map(propertyPath => propertyPath.get('value'))[0];
}
