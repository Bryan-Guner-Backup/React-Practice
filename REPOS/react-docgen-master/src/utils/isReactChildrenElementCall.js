/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import isReactModuleName from './isReactModuleName';
import match from './match';
import resolveToModule from './resolveToModule';
import type { Importer } from '../types';

/**
 * Returns true if the expression is a function call of the form
 * `React.Children.only(...)`.
 */
export default function isReactChildrenElementCall(
  path: NodePath,
  importer: Importer,
): boolean {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }

  if (!match(path.node, { callee: { property: { name: 'only' } } })) {
    return false;
  }

  const calleeObj = path.get('callee', 'object');
  const module = resolveToModule(calleeObj, importer);

  if (!match(calleeObj, { value: { property: { name: 'Children' } } })) {
    return false;
  }

  return Boolean(module && isReactModuleName(module));
}
