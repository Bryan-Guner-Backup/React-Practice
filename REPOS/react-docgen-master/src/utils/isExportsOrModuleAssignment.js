/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 */

import { namedTypes as t } from 'ast-types';
import * as expressionTo from './expressionTo';
import type { Importer } from '../types';

/**
 * Returns true if the expression is of form `exports.foo = ...;` or
 * `modules.exports = ...;`.
 */
export default function isExportsOrModuleAssignment(
  path: NodePath,
  importer: Importer,
): boolean {
  if (t.ExpressionStatement.check(path.node)) {
    path = path.get('expression');
  }
  if (
    !t.AssignmentExpression.check(path.node) ||
    !t.MemberExpression.check(path.node.left)
  ) {
    return false;
  }

  const exprArr = expressionTo.Array(path.get('left'), importer);
  return (
    (exprArr[0] === 'module' && exprArr[1] === 'exports') ||
    exprArr[0] === 'exports'
  );
}
