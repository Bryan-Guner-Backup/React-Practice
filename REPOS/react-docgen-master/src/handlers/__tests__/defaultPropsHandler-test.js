/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 */

jest.mock('../../Documentation');

import { parse, noopImporter, makeMockImporter } from '../../../tests/utils';

describe('defaultPropsHandler', () => {
  let documentation;
  let defaultPropsHandler;

  beforeEach(() => {
    documentation = new (require('../../Documentation'))();
    defaultPropsHandler = require('../defaultPropsHandler').default;
  });

  const mockImporter = makeMockImporter({
    getDefaultProps: parse(`
      import baz from 'baz';
      export default function() {
        return {
          foo: "bar",
          bar: 42,
          baz: baz,
          abc: {xyz: abc.def, 123: 42}
        };
      }
    `).get('body', 1, 'declaration'),

    baz: parse(`
      export default ["foo", "bar"];
    `).get('body', 0, 'declaration'),

    other: parse(`
      export default { bar: "foo" };
    `).get('body', 0, 'declaration'),

    defaultProps: parse(`
      export default {
        foo: "bar",
        bar: 42,
        baz: ["foo", "bar"],
        abc: {xyz: abc.def, 123: 42}
      };
    `).get('body', 0, 'declaration'),
  });

  describe('ObjectExpression', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              bar: 42,
              baz: ["foo", "bar"],
              abc: {xyz: abc.def, 123: 42}
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('can resolve declared functions', () => {
      const src = `
        function getDefaultProps() {
          return {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
        }
        ({
          getDefaultProps: getDefaultProps
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should find prop default values that are literals from imported functions', () => {
      const src = `
        import getDefaultProps from 'getDefaultProps';

        ({
          getDefaultProps: getDefaultProps
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('handles computed properties', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [bar]: 42,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('handles imported values assigned to computed properties', () => {
      const src = `
        import baz from 'baz';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [bar]: baz,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('ignores complex computed properties', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [() => {}]: 42,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('ignores imported values assigned to complex computed properties', () => {
      const src = `
        import baz from 'baz';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [() => {}]: baz,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              ...other,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              ...other,
            };
          }
        })
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassDeclaration with static defaultProps', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        class Foo {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported values assigned as default props', () => {
      const src = `
        import defaultProps from 'defaultProps';
        class Foo {
          static defaultProps = defaultProps;
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should resolve local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        class Foo {
          static defaultProps = {
            foo: "bar",
            ...other
          };
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        class Foo {
          static defaultProps = {
            foo: "bar",
            ...other
          };
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent';

        class Foo {
          static defaultProps = {
            foo: ImportedComponent,
          };
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('can resolve default props that are imported given a custom importer', () => {
      const src = `
        import baz from 'baz';

        class Foo {
          static defaultProps = {
            baz: baz,
          };
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassExpression with static defaultProps', () => {
    it('should find prop default values that are literals', () => {
      const src = `
        var Bar = class {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
      }`;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'declarations', 0, 'init'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported values assigned as default props', () => {
      const src = `
        import defaultProps from 'defaultProps';
        var Bar = class {
          static defaultProps = defaultProps;
        }
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'declarations', 0, 'init'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  it('should only consider Property nodes, not e.g. spread properties', () => {
    const src = `
      ({
        getDefaultProps: function() {
          return {
            ...Foo.bar,
            bar: 42,
          };
        }
      })
    `;
    const definition = parse(src).get('body', 0, 'expression');
    expect(() =>
      defaultPropsHandler(documentation, definition, noopImporter),
    ).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  it('can have an importer that resolves spread properties', () => {
    const src = `
      import Props from 'defaultProps';
      ({
        getDefaultProps: function() {
          return {
            ...Props.abc,
            bar: 42,
          };
        }
      })
    `;
    const definition = parse(src).get('body', 1, 'expression');
    expect(() =>
      defaultPropsHandler(documentation, definition, mockImporter),
    ).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  describe('Functional components with default params', () => {
    it('should find default props that are literals', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = {xyz: abc.def, 123: 42}
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('can use imported values as default props', () => {
      const src = `
        import baz from 'baz';
        ({
          bar = baz,
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should override with defaultProps if available', () => {
      const src = `
        var Foo = ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = 'test'
        }) => <div />
        Foo.defaultProps = { abc: {xyz: abc.def, 123: 42} };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'declarations', 0, 'init'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('overrides with imported defaultProps', () => {
      const src = `
        import other from 'other';
        var Foo = ({
          bar = 42,
        }) => <div />
        Foo.defaultProps = other;
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'declarations', 0, 'init'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };
        var Foo = (props) => <div />
        Foo.defaultProps = { foo: "bar", ...other };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'declarations', 0, 'init'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        var Foo = (props) => <div />
        Foo.defaultProps = { foo: "bar", ...other };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'declarations', 0, 'init'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should work with aliases', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc: defg = {xyz: abc.def, 123: 42}
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('allows imported defaults to be aliased', () => {
      const src = `
        import baz from 'baz';
        ({
          foo: bar = baz,
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent';

        ({
          foo = ImportedComponent,
        }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('should work with no defaults', () => {
      const src = `
        ({ foo }) => <div />
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 0, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('forwardRef', () => {
    it('resolves default props in the parameters', () => {
      const src = `
        import React from 'react';
        React.forwardRef(({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported default props in the parameters', () => {
      const src = `
        import baz from 'baz';
        import React from 'react';
        React.forwardRef(({ bar = baz }, ref) => <div ref={ref}>{bar}</div>);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 2, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves defaultProps', () => {
      const src = `
        import React from 'react';
        const Component = React.forwardRef(({ foo }, ref) => <div ref={ref}>{foo}</div>);
        Component.defaultProps = { foo: 'baz' };
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 1),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves imported defaultProps', () => {
      const src = `
        import other from 'other';
        import React from 'react';
        const Component = React.forwardRef(({ bar }, ref) => <div ref={ref}>{bar}</div>);
        Component.defaultProps = other;
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 2),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('resolves when the function is not inline', () => {
      const src = `
        import React from 'react';
        const ComponentImpl = ({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>;
        React.forwardRef(ComponentImpl);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 2, 'expression'),
        noopImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    it('also resolves imports when the function is not inline', () => {
      const src = `
        import baz from 'baz';
        import React from 'react';
        const ComponentImpl = ({ bar = baz }, ref) => <div ref={ref}>{bar}</div>;
        React.forwardRef(ComponentImpl);
      `;
      defaultPropsHandler(
        documentation,
        parse(src).get('body', 3, 'expression'),
        mockImporter,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });
});
