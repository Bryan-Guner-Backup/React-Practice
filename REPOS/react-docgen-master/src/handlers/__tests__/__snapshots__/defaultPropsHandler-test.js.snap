// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultPropsHandler ClassDeclaration with static defaultProps can resolve default props that are imported given a custom importer 1`] = `
Object {
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps resolves imported spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps resolves imported values assigned as default props 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are imported variables 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should resolve local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassExpression with static defaultProps resolves imported values assigned as default props 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassExpression with static defaultProps should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params allows imported defaults to be aliased 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params can use imported values as default props 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params overrides with imported defaultProps 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params resolves imported spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params resolves local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find default props that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find prop default values that are imported variables 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should override with defaultProps if available 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with aliases 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with no defaults 1`] = `Object {}`;

exports[`defaultPropsHandler ObjectExpression can resolve declared functions 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression handles computed properties 1`] = `
Object {
  "@computed#bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression handles imported values assigned to computed properties 1`] = `
Object {
  "@computed#bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression ignores complex computed properties 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression ignores imported values assigned to complex computed properties 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression resolves imported spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression resolves local spreads 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression should find prop default values that are literals 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression should find prop default values that are literals from imported functions 1`] = `
Object {
  "abc": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "baz": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"bar\\"",
    },
  },
}
`;

exports[`defaultPropsHandler can have an importer that resolves spread properties 1`] = `
Object {
  "123": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
  "xyz": Object {
    "defaultValue": Object {
      "computed": true,
      "value": "abc.def",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef also resolves imports when the function is not inline 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves default props in the parameters 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves defaultProps 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'baz'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves imported default props in the parameters 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "[\\"foo\\", \\"bar\\"]",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves imported defaultProps 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "\\"foo\\"",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves when the function is not inline 1`] = `
Object {
  "foo": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler should only consider Property nodes, not e.g. spread properties 1`] = `
Object {
  "bar": Object {
    "defaultValue": Object {
      "computed": false,
      "value": "42",
    },
  },
}
`;
